A88-xds-data-error-handling.md

```diff
@@ -0,0 +1,451 @@
+cases where the server is unable to send a resource that the client has
+subscribed to.
+
+### Related Proposals: 
+- [gRFC A53: Option for Ignoring xDS Resource Deletion][gRFC A53]
+- [gRFC A57: XdsClient Failure Mode Behavior][gRFC A57]
+- [gRFC A40: xDS Configuration Dump via Client Status Discovery Service][gRFC A40]
+- [gRFC A78: OTel Metrics for WRR, Pick First, and XdsClient][gRFC A78]
+- [gRFC A54: Restrict Control-Plane Status Codes][gRFC A54]
+- [xRFC TP3: Error Propagation][xRFC TP3]
+
+[gRFC A53]: A53-xds-ignore-resource-deletion.md
+[gRFC A57]: A57-xds-client-failure-mode-behavior.md
+[gRFC A40]: A40-csds-support.md
+[gRFC A78]: A78-grpc-metrics-wrr-pf-xds.md
+[gRFC A54]: A54-restrict-control-plane-status-codes.md
+[xRFC TP3]: https://github.com/cncf/xds/blob/main/proposals/TP3-xds-error-propagation.md
+
+## Proposal
+
+This proposal has the following parts:
+- changes in the XdsClient watcher APIs
+- handling data errors and transient errors
+- support for errors provided by the xDS server as per [xRFC TP3]
+- changes to the resource timer behavior
+- deprecating the "ignore_resource_deletion" server feature
+
+### Changes to XdsClient Watcher APIs
+
+The current `XdsClient` API has three methods:
+
+- `OnResourceChanged(ResourceType resource)`: Invoked whenever a new
+  version of the resource is received from the xDS server.
+
+- `OnError(Status status)`: Invoked for all transient errors and for some
+  data errors (NACKs).  The watcher is generally expected to ignore the
+  error if it already has a valid cached resource.
+
+- `OnResourceDoesNotExist()`: Invoked specifically for the does-not-exist
+  case.  The watcher is generally expected to stop using any previously
+  cached resource and put itself into a failing state.
+  - Note: The mechanism described in [gRFC A53] inhibits these
+    notifications for only the case where we already had a cached version
+    of the resource, which in the SotW protocol applies only to LDS
+    and CDS.  That mechanism does not affect the case where we do not
+    already have a cached version of the resource, which is where the
+    does-not-exist timer is used.
+
+These methods do not map well to the desired error handling behavior
+defined above.
+
+This proposal replaces those methods with the following two methods:
+
+- `OnResourceChanged(StatusOr<ResourceType> resource)`: Will be invoked
+  to notify the watcher of a new version of the resource received from
+  the xDS server or an error indicating the reason why the resource cannot
+  be obtained.  Note that if an error is passed to this method after the
+  watcher has previously seen a valid resource, the watcher is expected
+  to stop using that previously delivered resource.  In this case, the
+  XdsClient will remove the resource from its cache, so that CSDS will
+  not report a resource that the client is not actually using.
+
+- `OnAmbientError(Status status)`: Will be invoked to notify the watcher
```

Comment:
Naming the argument `error` would make it clearer.

Reply:
I think it's clear that a status is an error, so I don't feel the need to change that.